### ISSUE TITLE
Captcha auto-completion.

### ISSUE THREAD

**Issue by majortomomeganebula** [K1rakishou/Kuroba-Experimental#758]
I just found out about a new imageboard flutter app for android and iOS called [Chance](https://github.com/moffatman/chan/), which has auto-completes the captcha for you. 

is there any way to implement this feature in Kuroba-Exp as well? 

![image](https://user-images.githubusercontent.com/103362208/162614523-17c00891-c9e9-45e7-9aca-d853ca8eb9b9.png)

**Comment by K1rakishou** [comment]
Someone will have to implement it because I have no idea how this shit is supposed to work and don't really feel like studying the source code.

**Comment by Gugui322** [comment]
> Someone will have to implement it because I have no idea how this shit is supposed to work and don't really feel like studying the source code.

https://github.com/AUTOMATIC1111/4chan-captcha-solver

Maybe this can help you.

**Comment by samlux04** [comment]
This would be sick. I tried (chance). My shitposting got so much productive. Ten times faster for the least. After completing one captcha it doesn't ask another captcha. I can keep posting several times (10+) without captcha window. What fantastic app. The drawback it is fckn laggy.

**Comment by zn-a** [comment]
> Someone will have to implement it because I have no idea how this shit is supposed to work and don't really feel like studying the source code.

Mass reply extension has it as well
https://github.com/HamletDuFromage/4chan-mass-reply

**Comment by K1rakishou** [comment]
So I decided to actually figure out how this shit works and in the end managed to take the trained model from https://github.com/AUTOMATIC1111/4chan-captcha-solver repository and successfully use it in my app. It can auto adjust the slider (not the model, you can do that without it) + extract the text from the captcha image. The accuracy seems to be pretty good, too. So if everything goes smoothly from now on then I will release a separate app which, once installed, will be used by KurobaEx to auto-solve captcha (because tensorflow libraries take like 50 megabytes only for 1 architecture, so I can't embed them into KurobaEx).

I won't be able to "fix" the model if the captcha changes again because I have no idea how. So it all depends on the pretrained model. But it should work while the captcha remains the same.

**Comment by samlux04** [comment]
4 chin now using 6 letters captcha. does it work with that?
Also what I notice with chance app, it doesn't prompt for captcha quite often. I could post like 10 shitpost before it prompt for a captcha. It also seems to be able to verify the captcha before uoloading the attachments. So  it feels faster to post. My shiposting has been tremendously improved by 500%. The thing is chance app is kinda laggy. I checked with fps meter. It actually around 60fps. It's the weird animation or something that make it appears laggy.

**Comment by K1rakishou** [comment]
Yes, it solves both 5 and 6 letter variants. The speed is like ~1 second per captcha. But it's on an emulator without the gpu acceleration, so it should be slightly faster on the actual devices (not all of them, only the ones that support tensorflow gpu acceleration). However, the slowest part is the slider auto adjustment (takes like 70% of total time, 662ms out of 855ms in this particular case). There is also 1 second of additional delay applied in this video, so it seems slower than it should be.

```
2022-08-12 14:45:49.029 D/Helpers: combineBgWithFgWithBestDisorderInternal() bestOffset=-5, bestDisorder=0.22017443
2022-08-12 14:45:49.040 D/Helpers: combineBgWithFgWithBestDisorder() took 662.970900ms
2022-08-12 14:45:49.089 D/Solver: solve() using cpu with 4 threads
2022-08-12 14:45:49.241 D/Solver: solve() took 193.174700ms, results: [RecognizedSequence(sequence=H8VD8H, confidence=1.0)]
```

I think it should be fine.

https://user-images.githubusercontent.com/10332458/184347831-55ff6182-2e02-4446-8078-0a3346c21faa.mp4

However, be aware that the model itself is not ideal and sometimes it fails to recognize some captchas or the slider auto adjustment fails (which can be improved if someone comes up with a better algorithm, but not me for sure because I'm too dumb for this kind of stuff). In the latter case you will be able to adjust the slider manually and it will solve the captcha with that correction. In the former case you will have to reload captcha or solve it manually.

**Comment by K1rakishou** [comment]
>I could post like 10 shitpost before it prompt for a captcha.

This has never happened to me even in the browser. Sometimes I get the "verification not required" but 10 times in a row? Never. The app also can handle the free captcha. So I don't know how you got those results.

>It also seems to be able to verify the captcha before uoloading the attachments. 

I'm not aware of this. If you know something and it actually exists then I will add the support.

**Comment by samlux04** [comment]
> > I could post like 10 shitpost before it prompt for a captcha.
> 
> This has never happened to me even in the browser. Sometimes I get the "verification not required" but 10 times in a row? Never. The app also can handle the free captcha. So I don't know how you get those results.
> 
> > It also seems to be able to verify the captcha before uoloading the attachments.
> 
> I'm not aware of this. If you know something and it actually exists then I will add the support.

Yeah. 5 times in a row was more common. I think it's because I used mobile data with random ip, dunno.
I noticed when posting with an attachment it was faster with chance app before it asked for a captcha. With kuroba it's considerably slower. I could tell by the notification it upload the attachments before asking for captha. It took 5-10 sec if I have an attachment. With chance app I didn't notice this delay. It's like posting regular post without attachment.

**Comment by K1rakishou** [comment]
>I could tell by the notification it upload the attachments before asking for captha.

Are you by any chance using "Attempt to post without captcha" reply mode? Open the reply layout and then longtap the send button. And change it to "Solve captcha then post".

**Comment by samlux04** [comment]
> > I could tell by the notification it upload the attachments before asking for captha.
> 
> Are you by any chance using "Attempt to post without captcha" reply mode? Open the reply layout and then longtap the send button. And change it to "Solve captcha then post".

Yes, yes that's what I'm talking about. If I enable "attempt to post without captcha", it would try to upload first then asks for captcha. With it disable of course it would ask for a captcha every time. With chances there's no such option, in fact there's barely any options. it post without captcha just fine, and if there's any it would ask for it. But without much delay. That's what I'm curious about. Is it not uploading the attachments first?

**Comment by K1rakishou** [comment]
I don't think there is a way to check captcha first and then upload everything else. At least I don't know about it.
